tibanna_ffcommon/generic_qc_utils.py

```diff
@@ -1,7 +1,160 @@
-from .exceptions import (
-    GenericQcException
-)
-from typing import Any, List
+from .exceptions import GenericQcException
+from typing import Any, List, Union, Optional
+from pydantic import BaseModel, ConfigDict, ValidationError, RootModel
+from .misc_utils import LogicalExpressionParser
+
+# Tibanna interal QC flags
+PASS = "pass"
+WARN = "warn"
+FAIL = "fail"
+
+
+class QC_threshold(BaseModel):
+    id: str
+    metric: str
+    operator: str
+    pass_target: Union[float, int, str]
+    warn_target: Union[float, int, str]
+    use_as_qc_flag: Optional[bool] = None
+
+
+class QC_ruleset(BaseModel):
+    qc_thresholds: List[QC_threshold]
+    overall_quality_status_rule: str
+
+
+class QC_json_value(BaseModel):
+    key: str
+    value: Any
+
+    model_config = ConfigDict(extra="allow")
+
+
+class QC_json(BaseModel):
+    qc_values: List[QC_json_value]
+
+
+def validate_qc_ruleset(qc_ruleset) -> QC_ruleset:
+    """Validate the user supplied QC rule set again the schema and return a Pydantic model.
+
```

Comment:
This is a little odd naming I'd say... what you're really doing here is instantiating an object, and as part of that doing validation